llumnix/global_scheduler/migration_filter.py

```diff
@@ -0,0 +1,146 @@
+        filtered_dst_instance_infos = [info for info in instance_infos if all(cond(info) for cond in dst_filter_conditions)]
+
+        return filtered_src_instance_infos, filtered_dst_instance_infos
+
+class LoadConstrainedFilter(MigrationFilterPolicy):
+    def filter_src_condition(self, filter_config: MigrationFilterConfig,
+                             pair_migration_type: PairMigrationConstraints) -> Callable[[InstanceInfo], bool]:
+        return lambda instance_info: instance_info.num_killed_requests > 0 \
+            or instance_info.instance_load_migrate > filter_config.migrate_out_load_threshold
+
+    def filter_dst_condition(self, filter_config: MigrationFilterConfig,
+                             pair_migration_type: PairMigrationConstraints) -> Callable[[InstanceInfo], bool]:
+        return lambda instance_info: instance_info.num_killed_requests == 0 \
+            and instance_info.instance_load_migrate < filter_config.migrate_out_load_threshold
+
+class PddFilter(MigrationFilterPolicy):
+    INSTANCE_FILTER_RULES = {
+        PairMigrationConstraints.DECODING_2_DECODING: (InstanceType.DECODE, InstanceType.DECODE),
+        PairMigrationConstraints.PREFILL_2_DECODING: (InstanceType.PREFILL, InstanceType.DECODE),
+    }
+
+    def filter_src_condition(self, filter_config: MigrationFilterConfig,
+                             pair_migration_type: PairMigrationConstraints) -> Callable[[InstanceInfo], bool]:
+        src_type, _ = self.INSTANCE_FILTER_RULES[pair_migration_type]
+        instance_type_filter = lambda instance_info: instance_info.instance_type == src_type
+
+        if pair_migration_type == PairMigrationConstraints.DECODING_2_DECODING:
+            inner_policy = MigrationFilterPolicyFactory.get_policy('loadconstraint')
+            policy_filter = inner_policy.filter_src_condition(filter_config, pair_migration_type)
+        else:
+            policy_filter = lambda instance_info: True
+
+        return lambda instance_info: instance_type_filter(instance_info) and policy_filter(instance_info)
+
+    def filter_dst_condition(self, filter_config: MigrationFilterConfig,
+                             pair_migration_type: PairMigrationConstraints) -> Callable[[InstanceInfo], bool]:
+        _, dst_type = self.INSTANCE_FILTER_RULES[pair_migration_type]
+        instance_type_filter = lambda instance_info: instance_info.instance_type == dst_type
+
+        if pair_migration_type == PairMigrationConstraints.DECODING_2_DECODING:
+            inner_policy = MigrationFilterPolicyFactory.get_policy('loadconstraint')
+            policy_filter = inner_policy.filter_dst_condition(filter_config, pair_migration_type)
+        else:
+            policy_filter = lambda instance_info: instance_info.num_killed_requests == 0
+
+        return lambda instance_info: instance_type_filter(instance_info) and policy_filter(instance_info)
+
+class CustomFilter(MigrationFilterPolicy):
```

Comment:
CustomFilter is too general, I think SpecifiedFilter is enough, this can be a filter pre-registered in our system, and dynamically changing specified instance_ids outside.

Reply:
how about MigrationCustomFilter?

Reply:
get

Reply:
I think updating the corresponding filter_condition （using CustomFilter.set_filter_condition ）can achieve the effect you described.